File: web/src/main/java/no/nav/modiapersonoversikt/service/unleash/UnleashServiceImpl.java
package no.nav.modiapersonoversikt.service.unleash;

import io.getunleash.Unleash;
import io.getunleash.repository.ClientFeaturesResponse;
import io.getunleash.repository.FeatureToggleResponse;
import io.getunleash.repository.HttpFeatureFetcher;
import no.nav.common.health.selftest.SelfTestCheck;
import no.nav.modiapersonoversikt.infrastructure.ping.ConsumerPingable;
import org.slf4j.Logger;

import java.net.ConnectException;

import static io.getunleash.repository.FeatureToggleResponse.Status.UNAVAILABLE;
import static org.slf4j.LoggerFactory.getLogger;

public class UnleashServiceImpl implements UnleashService {

    private static final Logger log = getLogger(UnleashServiceImpl.class);

    private final Unleash defaultUnleash;
    private final HttpFeatureFetcher toggleFetcher;
    private final ConsumerPingable pingDelegate;

    String api;

    public UnleashServiceImpl(HttpFeatureFetcher toggleFetcher, Unleash defaultUnleash, String api) {
        this.toggleFetcher = toggleFetcher;
        this.defaultUnleash = defaultUnleash;
        this.api = api;

        this.pingDelegate = new ConsumerPingable("Unleash", false, () -> {
            ClientFeaturesResponse featuresResponse = this.toggleFetcher.fetchFeatures();
            FeatureToggleResponse.Status status = featuresResponse.getStatus();
            if (status.equals(UNAVAILABLE)) {
                throw new ConnectException("Ping mot Unleash på " + api + ". Ga status " + status);
            }
        });
    }

    @Override
    public boolean isEnabled(Feature feature) {
        return defaultUnleash.isEnabled(feature.getPropertyKey());
    }

    @Override
    public boolean isEnabled(String feature) {
        return defaultUnleash.isEnabled(feature);
    }

    @Override
    public SelfTestCheck ping() {
        return pingDelegate.ping();
    }
}

File: web/src/main/java/no/nav/modiapersonoversikt/service/unleash/UnleashServiceImpl.kt
package no.nav.modiapersonoversikt.service.unleash

import io.getunleash.Unleash
import io.getunleash.event.ClientFeaturesResponse
import io.getunleash.repository.HttpFeatureFetcher
import no.nav.common.health.selftest.SelfTestCheck
import no.nav.common.utils.fn.UnsafeRunnable
import no.nav.modiapersonoversikt.infrastructure.ping.ConsumerPingable

class UnleashServiceImpl(
    private val toggleFetcher: HttpFeatureFetcher,
    private val defaultUnleash: Unleash,
    var api: String?,
) : UnleashService {
    private val pingDelegate: ConsumerPingable =
        ConsumerPingable(
            "Unleash",
            false,
            UnsafeRunnable {
                val featuresResponse: ClientFeaturesResponse = this.toggleFetcher.fetchFeatures()
                val status: ClientFeaturesResponse.Status = featuresResponse.status
                if (status == ClientFeaturesResponse.Status.UNAVAILABLE) {
                    throw java.net.ConnectException("Ping mot Unleash på $api. Ga status $status")
                }
            },
        )

    override fun isEnabled(feature: no.nav.modiapersonoversikt.service.unleash.Feature): kotlin.Boolean =
        defaultUnleash.isEnabled(feature.propertyKey)

    override fun isEnabled(feature: kotlin.String): kotlin.Boolean = defaultUnleash.isEnabled(feature)

    override fun ping(): SelfTestCheck? = pingDelegate.ping()

    companion object {
        private val log: org.slf4j.Logger? = org.slf4j.LoggerFactory.getLogger(UnleashServiceImpl::class.java)
    }
}
